src/IO/H5/VolumeData.cpp

```diff
@@ -154,6 +162,308 @@ void append_element_extents_and_connectivity(
+    for (const auto& sorted_element_index : sorted_block_index) {
+      sizing_vector.push_back(property_to_sort[sorted_element_index]);
+    }
+    sorted_property.push_back(sizing_vector);
+  }
+
+  return sorted_property;
+}
+
+// Returns properties for each block
+template <size_t SpatialDim>
+std::tuple<size_t, size_t, std::array<int, SpatialDim>>
+compute_block_level_properties(
+    const std::vector<std::string>& block_grid_names,
+    const std::vector<std::vector<size_t>>& block_extents) {
+  size_t expected_connectivity_length = 0;
+  size_t expected_number_of_grid_points = 0;
+
+  for (const auto& extents : block_extents) {
+    size_t element_grid_points = 1;
+    size_t element_connectivity_length = 1;
+    for (size_t j = 0; j < SpatialDim; j++) {
+      element_grid_points *= extents[j];
+      element_connectivity_length *= extents[j] - 1;
+    }
+    // Connectivity that already exists
+    expected_connectivity_length += element_connectivity_length * 8;
+    // Used for reserving the length of block_logical_coords
+    expected_number_of_grid_points += element_grid_points;
+  }
+
+  std::string grid_name_string = block_grid_names[0];
+  std::array<int, SpatialDim> h_ref_array = {};
+  size_t h_ref_previous_start_position = 0;
+  size_t additional_connectivity_length = 1;
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    size_t h_ref_start_position =
+        grid_name_string.find('L', h_ref_previous_start_position + 1);
+    size_t h_ref_end_position =
+        grid_name_string.find('I', h_ref_start_position);
+    int h_ref = std::stoi(
+        grid_name_string.substr(h_ref_start_position + 1,
+                                h_ref_end_position - h_ref_start_position - 1));
+    h_ref_array[i] = h_ref;
+    additional_connectivity_length *= pow(2, h_ref + 1) - 1;
+    h_ref_previous_start_position = h_ref_start_position;
+  }
+
+  expected_connectivity_length +=
+      (additional_connectivity_length - block_extents.size()) * 8;
+
+  return std::tuple{expected_connectivity_length,
+                    expected_number_of_grid_points, h_ref_array};
+}
+
+// Generates the mesh
+template <size_t SpatialDim>
+Mesh<SpatialDim> generate_element_mesh(
+    const std::vector<std::string>& element_bases,
+    const std::vector<std::string>& element_quadratures,
+    const std::vector<size_t>& element_extents) {
+  std::array<Spectral::Basis, SpatialDim> basis_array = {};
+
+  std::array<Spectral::Quadrature, SpatialDim> quadrature_array = {};
+
+  std::array<size_t, SpatialDim> extents_array = {};
+
+  for (size_t i = 0; i < SpatialDim; i++) {
+    basis_array[i] = Spectral::to_basis(element_bases[i]);
+    quadrature_array[i] = Spectral::to_quadrature(element_quadratures[i]);
+    extents_array[i] = element_extents[i];
+  }
+
+  return Mesh<SpatialDim>{extents_array, basis_array, quadrature_array};
+}
+
+template <size_t SpatialDim>
+std::vector<std::array<double, SpatialDim>> generate_block_logical_coordinates(
```

Comment:
Could this somehow reuse some of the code used by `Domain/BlockLogicalCoordinates.hpp`? This function is not easy to follow

Reply:
That sounds right. Yep, I would comment the code more. I'm not sure about this, but it might make sense to have it be either a separate file or in `BlockLogicalCoordinates.?pp` but I wouldn't worry about that for now. The only reason I'm pondering that is in case you think there will be other places in the code that would need the same operation, but I can't think of anywhere right now.